pure.1
.TH Pure 1 "March 2008" "Pure Version 0.x"
.SH NAME
pure \- the Pure interpreter
.SH SYNOPSIS
\fBpure\fP [-h] [-i] [-n] [-v[\fIlevel\fP]] [\fIscript\fP ...] [-- \fIargs\fP ...]
.SH OPTIONS
.TP
.B -h
Print help message and exit.
.TP
.B -i
Force interactive mode (read commands from stdin).
.TP
.B -n
Suppress automatic inclusion of the prelude.
.TP
.B -v
Set verbosity level. See below for details.
.TP
.B --
Stop option processing and pass the remaining command line arguments in the
.B argv
variable.
.SH DESCRIPTION
Pure is a modern-style functional programming language based on term
rewriting. Pure programs are basically collections of equational rules used to
evaluate expressions in a symbolic fashion by reducing them to normal form. A
brief overview of the language can be found in the \fBPURE OVERVIEW\fP section
below. (In case you're wondering, the name ``Pure'' actually refers to the
adjective. But you can also write it as ``PURE'' and take this as a recursive
acronym for the ``Pure Universal Rewriting Engine''.)
.PP
.B pure
is the Pure interpreter. The interpreter has an LLVM backend which
JIT-compiles Pure programs to machine code, hence programs run blazingly fast
and interfacing to C modules is easy, while the interpreter still provides a
convenient, fully interactive environment for running Pure scripts and
evaluating expressions.
.PP
If any source scripts are specified on the command line, they are loaded and
executed, after which the interpreter exits. Otherwise the interpreter enters
the interactive read-eval-print loop. You can also use the
.B -i
option to enter the interactive loop (continue reading from stdin) even after
processing some source scripts. To exit the interpreter, just type the
.B quit
command or the end-of-file character (^D on Unix) at the beginning of the
command line.
.PP
When the interpreter is in interactive mode and reads from a tty, commands are
read using
.BR readline (3)
(providing completion for all commands listed in section
.B INTERACTIVE USAGE
below, as well as for global function and variable symbols) and, when exiting
the interpreter, the command history is stored in
.BR ~/.pure_history ,
from where it is restored the next time you run the interpreter.
.PP
Options and source files are processed in the order in which they are given on
the command line. Processing of options and source files ends when the
.B --
option is encountered. Any following parameters are passed to the executing
script by means of the global
.B argc
and
.B argv
variables. Moreover, the
.B version
variable is set to the Pure interpreter version.
.PP
If available, the prelude script
.B prelude.pure
is loaded by the interpreter prior to any other other definitions, unless the
.B -n
option is specified. The prelude as well as other source scripts specified
with a relative pathname are first searched for in the current directory and
then in the directory specified with the
.B PURELIB
environment variable. If the
.B PURELIB
variable is not set, a system-specific default is used.
.PP
The
.B -v
option is most useful for debugging the interpreter, or if you are interested
in the code your program gets compiled to. The
.I level
argument is optional; it defaults to 1. Six different levels are implemented
at this time (two more bits are reserved for future extensions). For most
purposes, only the first two levels will be useful for the average Pure
programmer; the remaining levels are most likely to be used by the Pure
interpreter developers.
.TP
.B 1 (0x1)
denotes echoing of parsed definitions and expressions;
.TP
.B 2 (0x2)
adds special annotations concerning local bindings (de Bruijn indices, subterm
paths; this can be helpful to debug tricky variable binding issues);
.TP
.B 4 (0x4)
adds abstract code snippets (matching automata etc.; you probably want to see
this only when working on the guts of the interpreter).
.TP
.B 8 (0x8)
dumps the ``real'' output code (LLVM assembler, which is as close to the
native machine code for your program as it gets; you \fIdefinitely\fP don't
want to see this unless you have to inspect the generated code for bugs or
performance issues).
.TP
.B 16 (0x10)
adds debugging messages from the
.BR bison (1)
parser; useful for debugging the parser.
.TP
.B 32 (0x20)
adds debugging messages from the
.BR flex (1)
lexer; useful for debugging the lexer.
.PP
These values can be or'ed together, and, for convenience, can be specified in
either decimal or hexadecimal. Thus 0xff always gives you full debugging
output (which isn't most likely be used by anyone but the Pure developers).
.PP
Note that the
.B -v
option is only applied \fIafter\fP the prelude has been loaded. If you want to
debug the prelude, use the
.B -n
option and specify the
.B prelude.pure
file explicitly on the command line. Alternatively, you can also use the
interactive
.B list
command (see the \fBINTERACTIVE USAGE\fP section below) to list definitions
along with additional debugging information.
.SH PURE OVERVIEW
.PP
Pure is a fairly simple language. Programs are simply collections of
equational rules defining functions, \fBlet\fP commands binding global
variables, and expressions to be evaluated. Here's a simple example, entered
interactively in the interpreter:
.sp
.nf
> // my first Pure example
> fact 1 = 1;
> fact n::int = n*fact (n-1) \fBif\fP n>1;
> \fBlet\fP x = fact 10; x;
3628800
.fi
.PP
The language is free-format (blanks are insignificant). As indicated,
definitions and expressions at the toplevel have to be terminated with a
semicolon. Comments have the same syntax as in C++ (using // for line-oriented
and /* ... */ for multiline comments; the latter may not be nested).
.PP
On the surface, Pure is quite similar to other modern functional languages
like Haskell and ML. But under the hood it is a much more dynamic and
reflective language, more akin to Lisp. In particular, Pure is dynamically
typed, so functions can be fully polymorphic and you can add to the definition
of an existing function at any time:
.sp
.nf
> fact 1.0 = 1.0;
> fact n::double = n*fact (n-1) \fBif\fP n>1;
> fact 10.0;
3628800.0
> fact 10;
3628800
.fi
.sp
Also, due to its term rewriting semantics, Pure can do symbolic evaluations:
.sp
.nf
> square x = x*x;
> square (a+b);
(a+b)*(a+b)
.fi
.PP
The Pure language provides built-in support for machine integers (32 bit),
bigints (implemented using GMP), floating point values (double precision
IEEE), character strings (UTF-8 encoded) and generic C pointers (these don't
have a syntactic representation in Pure, though, so they need to be created
with external C functions). Truth values are encoded as machine integers (as
you might expect, zero denotes ``false'' and any non-zero value ``true'').
.PP
Expressions consist of the following elements:
.TP
.B Constants: \fR4711, 1.2e-3, \(dqHello,\ world!\en\(dq
The usual C'ish notations for integers (decimal, hexadecimal, octal), floating
point values and double-quoted strings are all provided. Actually, in Pure
character escapes have a more flexible syntax borrowed from the author's Q
language, which provides notations to specify any Unicode character. In
particular, the notation
.BR \e\fIn\fP ,
where \fIn\fP is an integer written in decimal, hexadecimal or octal notation,
denotes the Unicode character (code point) #\fIn\fP. Since these escapes may
consist of a varying number of digits, parentheses may be used for
disambiguation purposes; thus, e.g.
.B \(dq\e(123)4\(dq
denotes character #123 followed by the character `4'. The usual C-like escapes
for special non-printable characters such as
.B \en
are also supported. Moreover, you can use symbolic character escapes of the
form
.BR \e&\fIname\fP; ,
where \fIname\fP is any of the XML single character entity names specified in
the ``XML Entity definitions for Characters'', see
.IR http://www.w3.org/TR/xml-entity-names/ .
.TP
.B Function and variable symbols: \fRfoo, foo_bar, BAR, bar2
These consist of the usual sequence of ASCII letters (including the
underscore) and digits, starting with a letter. Case is significant, but it
doesn't carry any meaning (that's in contrast to languages like Prolog and Q,
where variables must be capitalized). Pure simply distinguishes function and
variable symbols on the left-hand side of an equation by the ``head =
function'' rule: Any symbol which occurs as the head symbol of a function
application is a function symbol, all other symbols are variables -- except
symbols explicitly declared as ``constant'' a.k.a.
.B nullary
symbols, see below. Another important thing to know is that in Pure, keeping
with the tradition of term rewriting, there's no distinction between
``defined'' and ``constructor'' function symbols; any function symbol can also
act as a constructor if it happens to occur in a normal form term.
.TP
.B Operator and constant symbols: \fRx+y, x==y, \fBnot\fP\ x
As indicated, these take the form of an identifier or a sequence of ASCII
punctuation symbols, as defined in the source using corresponding
\fBprefix\fP, \fBpostfix\fP and \fBinfix\fP declarations, which are discussed
in section DECLARATIONS. Enclosing an operator in parentheses, such as (+) or
(\fBnot\fP), turns it into an ordinary function symbol. Symbols can also be
defined as \fBnullary\fP to denote special constant symbols. See the prelude
for examples.
.TP
.B Lists and tuples: \fR[x,y,z], x..y, x:xs, x,y,z
The necessary constructors to build lists and tuples are actually defined in
the prelude: `[]' and `()' are the empty list and tuple, `:' produces list
``conses'', and `,' produces ``pairs''. As indicated, Pure provides the usual
syntactic sugar for list values in brackets, such as [x,y,z], which is exactly
the same as x:y:z:[]. Moreover, the prelude also provides an infix `..' 
operator to denote arithmetic sequences such as 1..10 or 1.0,1.2..3.0. Pure's
tuples are a bit unusual, however, in that they abide by the mathematical
rules: They are constructed by just ``paring'' things using the `,' operator,
for which the empty tuple acts as a neutral element (i.e., (),x is just x, as
is x,()). The pairing operator is associative, which implies that tuples are
completely flat (i.e., x,(y,z) is just x,y,z, as is (x,y),z). This means that
there are no nested tuples (tuples of tuples), if you need such constructs
then you should use lists instead. Also note that the parentheses are
\fInot\fP part of the tuple syntax in Pure, although you \fIcan\fP use
parentheses, just as with any other expression, for the usual purpose of
grouping expressions and overriding default precedences and associativity.
.TP
.B List comprehensions: \fR[x,y; x = 1..n; y = 1..m; x<y]
Pure also has list comprehensions which generate lists from an expression and
one or more ``generator'' and ``filter'' clauses (the former bind a pattern to
values drawn from a list, the latter are just predicates determining which
generated elements should actually be added to the output list). List
comprehensions are in fact syntactic sugar for a combination of nested
lambdas, conditional expressions and ``catmaps'' (a list operation which
combines list concatenation and mapping a function over a list, defined in the
prelude), but they are often much easier to write.
.TP
.B Function applications: \fRfoo\ x\ y\ z
As in other modern FPLs, these are written simply as juxtaposition (i.e., in
``curried'' form) and associate to the left. Operator applications are written
using prefix, postfix or infix notation, as the declaration of the operator
demands, but are just ordinary function applications in disguise. E.g., x+y is
exactly the same as (+) x y.
.TP
.B Conditional expressions: if\fR\ x\ \fBthen\fR\ y\ \fBelse\fR\ z
Evaluates to y or z depending on whether x is ``true'' (i.e., a nonzero
integer).
.TP
.B Lambdas: \fR\ex\ ->\ y
These work pretty much like in Haskell. More than one variable may be bound
(e.g, \ex\ y\ ->\ x*y), which is equivalent to a nested lambda
(\ex\ ->\ \ey\ ->\ x*y). Pure also fully supports pattern-matching lambda
abstractions which match a pattern against the lambda argument and bind
multiple lambda variables in one go, such as \e(x,y)\ ->\ x*y.
.TP
.B Case expressions: case\fR\ x\ \fBof\fR\ \fIrule\fR;\ ...\ \fBend
Matches an expression, discriminating over a number of different patterns;
similar to the Haskell \fBcase\fP construct.
.TP
.B When expressions: \fRx\ \fBwhen\fR\ \fIrule\fR;\ ...\ \fBend
An alternative way to bind local variables by matching a collection of subject
terms against corresponding patterns. Similar to Aardappel's \fBwhen\fP
construct, but Pure allows more than one definition. Note that multiple
definitions in a \fBwhen\fP clause are processed from left to right, so that
later definitions may refer to the variables in earlier ones. In fact, a
\fBwhen\fP expression with multiple definitions is treated like several
nested \fBwhen\fP expressions, with the first binding being the ``outermost''
one.
.TP
.B With expressions: \fRx\ \fBwith\fR\ \fIrule\fR;\ ...\ \fBend\fR
Defines local functions. Like Haskell's \fBwhere\fP construct, but can be used
anywhere inside an expression (just like Aardappel's \fBwhere\fP, but Pure
uses the keyword \fBwith\fP which better lines up with \fBcase\fP and
\fBwhen\fP). Also note that while Haskell lets you do \fIboth\fP function
definitions and ``pattern bindings'' in its \fBwhere\fP clauses, in Pure you
have to use \fBwith\fP for the former and \fBwhen\fP for the latter. This is
necessary because Pure, in contrast to Haskell, does not distinguish between
defined functions and constructors and thus there is no magic to figure out
whether an equation is meant as a function definition or a pattern binding.
.PP
At the toplevel, a Pure program basically consists of rules a.k.a. equations
defining functions, variable definitions a.k.a. global ``pattern bindings'',
and expressions to be evaluated.
.TP
.B Rules: \fIlhs\fR = \fIrhs\fR;
The basic form can also be augmented with a condition \fBif\ \fIguard\fR
tacked on to the end of the rule (which restricts the applicability of the
rule to the case that the guard evaluates to a nonzero integer), or the
keyword
.B otherwise
denoting an empty guard which is always true (this is nothing but syntactic
sugar useful to point out the ``default'' case of a definition; the
interpreter just treats
.B otherwise
as a comment, so it can always be omitted). Moreover, the left-hand side can
be omitted if it is the same as for the previous rule. This provides a
convenient means to write out a collection of equations for the same left-hand
side which discriminates over different conditions:
.sp
.nf
\fIlhs\fR       = \fIrhs\fB if \fIguard\fR;
          = \fIrhs\fB if \fIguard\fR;
          ...
          = \fIrhs\fB otherwise\fR;
.fi
.sp
Rules are used to define functions at the toplevel and in \fBwith\fP
expressions, as well as inside \fBcase\fP and \fBwhen\fP expressions for the
purpose of performing pattern bindings (however, for obvious reasons the forms
without a left-hand side or including a guard are not permitted in \fBwhen\fP
expressions). When matching against a function call or the subject term in a
\fBcase\fP expression, the rules are always considered in the order in which
they are written, and the first matching rule (whose guard evaluates to a
nonzero value, if applicable) is picked. (Again, the \fBwhen\fP construct is
treated differently, because each rule is actually a separate pattern
binding.)
.sp
In any case, the left-hand side pattern must not contain repeated variables
(i.e., rules must be ``left-linear''), except for the ``anonymous'' variable
`_' which matches an arbitrary value without binding a variable
symbol. Moreover, a left-hand side variable may be followed by one of the
special type tags \fB::int\fP, \fB::bigint\fP, \fB::double\fP, \fB::string\fP,
to indicate that it can only match a constant value of the corresponding
built-in type. (This is useful if you want to write rules matching \fIany\fP
object of one of these types; note that there is no way to write out all
``constructors'' for the built-in types, as there are infinitely many.)
.TP
.B Global variable bindings: let\fR \fIlhs\fR = \fIrhs\fR;
This binds every variable in the left-hand side pattern to the corresponding
subterm of the evaluated right-hand side.
.TP
.B Toplevel expressions: \fIexpr\fR;
A singleton expression at the toplevel, terminated with a semicolon, simply
causes the given value to be evaluated (and the result to be printed, when
running in interactive mode).
.PP
For instance, here are two more function definitions showing most of these
elements in action:
.sp
.nf
fact n  = n*fact (n-1) \fBif\fP n>0;
        = 1 \fBotherwise\fP;

fib n   = a  \fBwhen\fP a, b   = fibs n \fBend\fP
             \fBwith\fP fibs n = 0, 1 \fBif\fP n<=0;
                         = \fBcase\fP fibs (n-1) \fBof\fP
                             a, b = b, a+b;
                           \fBend\fP;
             \fBend\fP;

\fBlet\fP facts = map fact (1..10); \fBlet\fP fibs = map fib (1..100);
facts; fibs;
.fi
.PP
And here's a little list comprehension example: Erathosthenes' classical prime
sieve.
.sp
.nf
primes n        = sieve (2..n) \fBwith\fP
  sieve []      = [];
  sieve (p:qs)  = p : sieve [q; q = qs; q mod p];
\fBend\fP;
.fi
.sp
For instance:
.sp
.nf
> primes 100;
[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97]
.fi
.PP
If you dare, you can actually have a look at the catmap-lambda-if-then-else
expression the comprehension expanded to:
.sp
.nf
> list primes
primes n = sieve (2..n) with sieve [] = []; sieve (p:qs) = p:sieve
(catmap (\eq -> if q mod p then [q] else []) qs) end;
.fi
.PP
List comprehensions are also a useful device to organize backtracking
searches. For instance, here's an algorithm for the n queens problem, which
returns the list of all placements of n queens on an n x n board (encoded as
lists of n pairs (i,j) with i = 1..n), so that no two queens hold each other
in check.
.sp
.nf
queens n        = search n 1 [] \fBwith\fP
  search n i p  = [reverse p] \fBif\fP i>n;
                = cat [search n (i+1) ((i,j):p); j = 1..n; safe (i,j) p];
  safe (i,j) p  = not any (check (i,j)) p;
  check (i1,j1) (i2,j2)
                = i1==i2 || j1==j2 || i1+j1==i2+j2 || i1-j1==i2-j2;
\fBend\fP;
.fi
.SH EXCEPTION HANDLING
Pure also offers a useful exception handling facility. To raise an exception,
you just invoke the built-in function
.B throw
with the value to be thrown as the argument. To catch an exception, you use
the built-in special form
.B catch
with the exception handler (a function to be applied to the exception value)
as the first and the expression to be evaluated as the second argument. For
instance:
.sp
.nf
> catch error (throw hello_world);
error hello_world
.fi
.PP
Exceptions are also generated by the runtime system if the program runs out of
stack space, when a guard does not evaluate to a truth value, and when the
subject term fails to match the pattern in a pattern-matching lambda
abstraction, or a \fBlet\fP, \fBcase\fP or \fBwhen\fP construct. These types
of exceptions are reported using the symbols
.BR stack_fault ,
.B failed_cond
and
.BR failed_match ,
respectively, which are declared as constant symbols in the standard
prelude. You can use
.B catch
to handle these kinds of exceptions just like any other. For instance:
.sp
.nf
> fact n = if n>0 then n*fact(n-1) else 1;
> catch error (fact foo);
error failed_cond
> catch error (fact 100000);
error stack_fault
.fi
.PP
(You'll only get the latter kind of exception if the interpreter does stack
checks, see the discussion of the
.B PURE_STACK
environment variable in the CAVEATS AND NOTES section.)
.PP
Note that unhandled exceptions are reported by the interpreter with a
corresponding error message:
.sp
.nf
> fact foo;
<stdin>:2.0-7: unhandled exception 'failed_cond' while evaluating 'fact foo'
.fi
.PP
Exceptions can also be used to implement non-local value returns. For
instance, here's a variation of our n queens algorithm which only returns the
first solution. Note the use of
.B throw
in the recursive search routine to bail out with a solution as soon as we
found one. The value thrown there is caught in the main routine. If no value
gets thrown, the function regularly returns with () to indicate that there is
no solution.
.sp
.nf
queens1 n       = catch reverse (search n 1 []) \fBwith\fP
  search n i p  = throw p \fBif\fP i>n;
                = void [search n (i+1) ((i,j):p); j = 1..n; safe (i,j) p];
  safe (i,j) p  = not any (check (i,j)) p;
  check (i1,j1) (i2,j2)
                = i1==i2 || j1==j2 || i1+j1==i2+j2 || i1-j1==i2-j2;
\fBend\fP;
.fi
.PP
E.g., let's compute a solution for a standard 8x8 board:
.sp
.nf
> queens 8;
(1,1):(2,5):(3,8):(4,6):(5,3):(6,7):(7,2):(8,4):[]
.fi
.SH DECLARATIONS
As you probably noticed, Pure is very terse. That's because, in contrast to
hopelessly verbose languages like Java, you don't declare much stuff in Pure,
you just define it and be done with it. Usually, all necessary information
about the defined symbols is inferred automatically. However, there are a few
toplevel constructs which let you declare special symbol attributes and manage
programs consisting of several source modules. These are: operator and
constant symbol declarations,
.B extern
declarations for external C functions (described in the next section), and
.B using
clauses which provide a simple include file mechanism.
.TP
.B Operator and constant declarations: infix \fIlevel\fP \fIop\fR ...;
Ten different precedence levels are available for user-defined operators,
numbered 0 (lowest) thru 9 (highest). On each precedence level, you can
declare (in order of increasing precedence)
.BR infix " (binary non-associative),"
.BR infixl " (binary left-associative),"
.BR infixr " (binary right-associative),"
.BR prefix " (unary prefix) and"
.BR postfix " (unary postfix)"
operators. For instance:
.sp
.nf
\fBinfixl\fP 6 + - ;
\fBinfixl\fP 7 * / div mod ;
.fi
.sp
Moreover, constant symbols are introduced using a declaration of
the form:
.sp
.nf
\fBnullary \fIsymbol\fR ...;
.fi
.sp
Examples for all of these can be found in the prelude which declares a bunch
of standard (arithmetic, relational, logical) operator symbols as well as the
list and pair constructors `:' and `,' and the constant symbols `[]' and `()'
denoting the empty list and tuple, respectively.
.TP
.B Using clause: using \fIname\fR ...;
Causes each given script to be included, at the position of the
.B using
clause, but only if the script was not included already. The script name can
be specified either as a string denoting the proper filename (possibly
including path and/or filename extension), or as an identifier. In the latter
case, the
.B .pure
filename extension is added automatically. In both cases, the script is
searched for in the current directory and the directory named by the
.B PURELIB
environment variable. (The
.B using
clause also has an alternative form which allows dynamic libraries to be
loaded, this will be discussed in the following section.)
.SH C INTERFACE
Accessing C functions from Pure programs is dead simple. You just need an
.B extern
declaration of the function, which is a simplified kind of C prototype. The
function can then be called in Pure just like any other. For instance, the
following commands, entered interactively in the interpreter, let you use the
.B sin
function from the C library (of course you could just as well put the
.B extern
declaration into a script):
.sp
.nf
> extern double sin(double);
> sin 0.3;
0.29552020666134
.fi
.sp
For clarity, the parameter types can also be annotated with parameter names,
e.g.:
.sp
.nf
extern double sin(double x);
.fi
.sp
Parameter names in prototypes only serve informational purposes and are for
the human reader; they are effectively treated as comments by the compiler.
.PP
The interpreter makes sure that the parameters in a call match; if not, the
call is treated as a normal form expression. The range of supported C types is
a bit limited right now (void, bool, char, int, double, as well as arbitrary
pointer types, i.e.: void*, char*, etc.), but in practice these should cover
most kinds of calls that need to be done when interfacing to C libraries. Note
that char* is for string arguments and return values which need translation
between Pure's internal utf-8 representation and the system encoding, while
void* is for any generic kind of pointer (including strings, which are
\fInot\fP translated when passed/returned as void*). Any other kind of pointer
(except expr*, see below), is effectively treated as void* right now, although
in a future version the interpreter may keep track of the type names for the
purpose of checking parameter types.
.PP
The expr* pointer type is special; it indicates a Pure expression parameter or
return value which is just passed through unchanged. All other types of values
have to be ``unboxed'' when they are passed as arguments (i.e., from Pure to
C) and ``boxed'' again when they are returned as function results (from C to
Pure). All of this is handled by the runtime system in a transparent way, of
course.
.PP
It is even possible to augment an external C function with ordinary Pure
equations, but in this case you have to make sure that the
.B extern
declaration of the function comes first. For instance, we might want to extend
our imported
.B sin
function with a rule to handle integers:
.sp
.nf
> sin 0;
sin 0
> sin x::int = sin (double x);
> sin 0;
0.0
.fi
.PP
Sometimes it is preferable to replace a C function with a wrapper function
written in Pure. In such a case you can specify an \fIalias\fP under which the
original C function is known to the Pure program, so that you can still call
the C function from the wrapper. An alias is introduced by terminating the
.B extern
declaration with a clause of the form ``= \fIalias\fP''. For instance:
.sp
.nf
> extern double sin(double) = c_sin;
> sin x::double = c_sin x;
> sin x::int = c_sin (double x);
> sin 0.3; sin 0;
0.29552020666134
0.0
.fi
.PP
External C functions are resolved by the LLVM runtime, which first looks for
the symbol in the interpreter executable. Since the interpreter links in its
own runtime support as well as all of the standard C library, these functions
are ready to be used in Pure programs. Other functions can be made available
by including them in the runtime, or by linking the interpreter against the
corresponding modules. Or, better yet, you can just ``dlopen'' shared
libraries at runtime with a special form of the
.B using
clause:
.sp
.nf
\fBusing\fP "lib:\fIlibname\fR[.\fIext\fP]";
.fi
.sp
For instance, if you want to call the GMP functions directly from Pure:
.sp
.nf
\fBusing\fP "lib:libgmp";
.fi
.sp
After this declaration the GMP functions will be ready to be imported into
your Pure program by means of corresponding
.B extern
declarations.
.PP
Shared libraries opened with \fBusing\fP clauses are searched for on the usual
system linker path (\fBLD_LIBRARY_PATH\fP on Linux). The necessary filename
suffix (\fB.so\fP on Linux) will also be supplied automatically. You can also
specify a full pathname for the library if you prefer that. If a library file

trunk/ChangeLog
2008-04-28  Albert Graef  <Dr.Graef@t-online.de>

	* examples/symbolic.pure: Add symbolic evaluation example. This is
	just a straightforward port of some Q examples.

	* runtime.cc: Add support for advisory stack checks.

	* matcher.cc: Bugfixes.

2008-04-27  Albert Graef  <Dr.Graef@t-online.de>

	* lib/string.pure: Added split and join functions.

	* pure.1, examples/hello.pure: Overhaul n queens example, added
	quicksort and binary search tree examples.

	* lib/prelude.pure: Added void and curry/uncurry combinators, do,
	zipdo.

	* interpreter.cc, pure.cc, runtime.cc: Make SIGINT generate a
	useful exception value.

	* pure.cc: Add completion for global function and variable
	symbols.

2008-04-22  Albert Graef  <Dr.Graef@t-online.de>

	* Got a working interpreter, at last. There's still lots of stuff
	to do (see the TODO file), but the interpreter should now be
	usable as it is.

2008-03-27  Albert Graef  <Dr.Graef@t-online.de>

	* ChangeLog started



pure/trunk/pure.cc

/* The Pure main program. This is currently rather simplistic. See the README
   file and the man page for details. */

#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <locale.h>
#include <signal.h>
#include <readline/readline.h>
#include <readline/history.h>
#include "interpreter.hh"
#include "runtime.h"
#include <llvm/Target/TargetOptions.h>

using namespace std;

#ifndef VERSION
#define VERSION "0.0"
#endif
#ifndef PURELIB
#define PURELIB "/usr/local/lib/pure"
#endif

#define COPYRIGHT "Copyright (c) 2008 by Albert Graef"
#define USAGE \
"Usage: pure [-h] [-i] [-n] [-v[level]] [script ...] [-- args ...]\n\
-h: Print this message and exit.\n\
-i: Force interactive mode (read commands from stdin).\n\
-n: Suppress automatic inclusion of the prelude.\n\
-v: Set verbosity level (useful for debugging purposes).\n\
--: Stop option processing, pass remaining args in argv variable.\n\
Environment:\n\
PURELIB:    Directory to search for source scripts including the prelude.\n\
PURE_PS:    Command prompt to be used in the interactive command loop.\n\
PURE_STACK: Maximum stack size in kilobytes (default: 0 = unlimited).\n"
#define LICENSE "This program is free software distributed under the GNU Public License\n(GPL V3 or later). Please see the COPYING file for details.\n"

const char *commands[] = {
  "cd", "clear", "extern", "help", "infix", "infixl", "infixr", "let", "list",
  "ls", "nullary", "override", "postfix", "prefix", "pwd", "quit", "run",
  "save", "stats", "underride", "using", 0
};

/* Generator functions for command completion. */

static char *
command_generator(const char *text, int state)
{
  static int list_index, len;
  static env::iterator it, end;
  const char *name;

  /* New match. */
  if (!state) {
    list_index = 0;
    assert(interpreter::g_interp);
    it = interpreter::g_interp->environ.begin();
    end = interpreter::g_interp->environ.end();
    len = strlen(text);
  }

  /* Return the next name which partially matches from the
     command list. */
  while ((name = commands[list_index])) {
    list_index++;
    if (strncmp(name, text, len) == 0)
      return strdup(name);
  }

  /* Return the next name which partially matches from the
     symbol list. */
  while (it != end) {
    assert(it->first > 0);
    symbol& sym = interpreter::g_interp->symtab.sym(it->first);
    it++;
    if (strncmp(sym.s.c_str(), text, len) == 0)
      return strdup(sym.s.c_str());
  }

  /* If no names matched, then return NULL. */
  return 0;
}

static char *
symbol_generator(const char *text, int state)
{
  static int len;
  static env::iterator it, end;

  /* New match. */
  if (!state) {
    assert(interpreter::g_interp);
    it = interpreter::g_interp->environ.begin();
    end = interpreter::g_interp->environ.end();
    len = strlen(text);
  }

  /* Return the next name which partially matches from the
     symbol list. */
  while (it != end) {
    assert(it->first > 0);
    symbol& sym = interpreter::g_interp->symtab.sym(it->first);
    it++;
    if (strncmp(sym.s.c_str(), text, len) == 0)
      return strdup(sym.s.c_str());
  }

  /* If no names matched, then return NULL. */
  return 0;
}

static char **
pure_completion(const char *text, int start, int end)
{
  char **matches;

  matches = (char **)NULL;

  /* If this word is at the start of the line, then it is a command to
     complete. Otherwise it is a global function or variable symbol, or the
     name of a file in the current directory. */
  if (start == 0)
    matches = rl_completion_matches(text, command_generator);
  else
    matches = rl_completion_matches(text, symbol_generator);

  return matches;
}

static void sigint_handler(int sig)
{
  interpreter::brkflag = sig;
}

static const char *histfile = 0;

static void exit_handler()
{
  if (histfile) write_history(histfile);
}

int
main(int argc, char *argv[])
{
  char base;
  interpreter interp;
  int count = 0;
  bool force_interactive = false, want_prelude = true, have_prelude = false;
  // This is used in advisory stack checks.
  interpreter::baseptr = &base;
  // make sure that SIGPIPE is ignored
  signal(SIGPIPE, SIG_IGN);
  /* Set up a SIGINT handler which throws a Pure exception, so that we safely
     return to the interpreter's interactive command line when the user
     interrupts a computation. */
  signal(SIGINT, sigint_handler);
  // set up an exit function which saves the history if needed
  atexit(exit_handler);
  // set the system locale
  setlocale(LC_ALL, "");
  // get some settings from the environment
  const char *env;
  if ((env = getenv("HOME")))
    interp.histfile = string(env)+"/.pure_history";
  if ((env = getenv("PURE_PS")))
    interp.ps = string(env);
  if ((env = getenv("PURE_STACK"))) {
    char *end;
    size_t n = strtoul(env, &end, 0);
    if (!*end) interpreter::stackmax = n*1024;
  }
  if ((env = getenv("PURELIB")))
    interp.lib = string(env)+"/";
  else
    interp.lib = string(PURELIB)+"/";
  string prelude = interp.lib+string("prelude.pure");
#if USE_FASTCC
  // This global option is needed to get tail call optimization (you'll also
  // need to have USE_FASTCC in interpreter.hh enabled).
  llvm::PerformTailCallOpt = true;
#endif
  // scan the command line options
  const string prog = *argv;
  list<string> myargs;
  for (char **args = ++argv; *args; ++args)
    if (*args == string("-h")) {
      cout << "Pure " << VERSION << " " << COPYRIGHT << endl << USAGE;
      return 0;
    } else if (*args == string("-i"))
      force_interactive = true;
    else if (*args == string("-n"))
      want_prelude = false;
    else if (string(*args).substr(0,2) == "-v") {
      string s = string(*args).substr(2);
      if (s.empty()) continue;
      char *end;
      strtoul(s.c_str(), &end, 0);
      if (*end) {
	interp.error(prog + ": invalid option " + *args);
	return 1;
      }
    } else if (*args == string("--")) {
      while (*++args) myargs.push_back(*args);
      break;
    } else if (**args == '-') {
      interp.error(prog + ": invalid option " + *args);
      return 1;
    }
  interp.init_sys_vars(VERSION, myargs);
  if (want_prelude) {
    // load the prelude if we can find it
    FILE *fp = fopen("prelude.pure", "r");
    if (fp)
      prelude = "prelude.pure";
    else
      // try again in the PURELIB directory
      fp = fopen(prelude.c_str(), "r");
    if (fp) {
      fclose(fp);
      have_prelude = true;
      interp.run(prelude);
      interp.compile();
    }
  }
  // load scripts specified on the command line
  for (; *argv; ++argv)
    if (string(*argv).substr(0,2) == "-v") {
      uint8_t level = 1;
      string s = string(*argv).substr(2);
      if (!s.empty()) level = (uint8_t)strtoul(s.c_str(), 0, 0);
      interp.verbose = level;
    } else if (*argv == string("--"))
      break;
    else if (**argv == '-')
      ;
    else if (**argv) {
      if (count++ == 0) interp.modname = *argv;
      interp.run(*argv);
    }
  if (count > 0 && !force_interactive) {
    if (interp.verbose&verbosity::dump) interp.compile();
    return 0;
  }
  interp.symtab.init_builtins();
  // enter the interactive command loop
  interp.interactive = true;
  if (isatty(fileno(stdin))) {
    // connected to a terminal, print sign-on and initialize readline
    cout << "Pure " << VERSION << " " << COPYRIGHT << endl << LICENSE;
    if (have_prelude)
      cout << "Loaded prelude from " << prelude << ".\n\n";
    rl_readline_name = "Pure";
    rl_attempted_completion_function = pure_completion;
    using_history();
    read_history(interp.histfile.c_str());
    stifle_history(600);
    histfile = strdup(interp.histfile.c_str());
    interp.compile();
    interp.ttymode = true;
  }
  interp.temp = 1;
  interp.run("");
  if (interp.ttymode) cout << endl;
  return 0;
}
